qa/tasks/quiescer.py

```diff
@@ -0,0 +1,269 @@
+"""
+Thrash mds by simulating failures
+"""
+import logging
+import contextlib
+
+from teuthology import misc as teuthology
+
+from tasks import ceph_manager
+from tasks.cephfs.filesystem import MDSCluster, Filesystem
+from tasks.thrasher import ThrasherGreenlet
+
+import random
+import math
+import errno
+import json
+import time
+
+from io import StringIO
+
+log = logging.getLogger(__name__)
+
+class Quiescer(ThrasherGreenlet):
+    """
+    The Quiescer does periodic quiescing of the configured paths, by default - the root '/'.
+
+    quiesce_timeout: maximum time in seconds to wait for the quiesce to succeed
+    quiesce_factor: the percentage of time we allow the system to stay quiesced
+    min_quiesce: the minimum pause time in seconds
+    max_quiesce: the maximum pause time in seconds
+    initial_delay: the time in seconds before the first quiesce
+    """
+
+    MAX_QUIESCE_FACTOR = 0.5    # 50%
+    MIN_QUIESCE_FACTOR = 0.005  # 0.5%
+
+    def __init__(self, fs, quiesce_timeout=30, quiesce_factor=0.1, min_quiesce=10, max_quiesce=60, initial_delay=120, seed=None, **unused_kwargs):
+        super(Quiescer, self).__init__()
+
+        self.logger = log.getChild('fs.[{f}]'.format(f=fs.name))
+        self.fs = fs
+        self.name = 'quiescer.fs.[{f}]'.format(f=fs.name)
+        self.rnd = random.Random()
+        self.rnd.seed(seed)
+
+        self.quiesce_timeout = quiesce_timeout
+
+        if (quiesce_factor > self.MAX_QUIESCE_FACTOR):
+            self.logger.warn("Capping the quiesce factor at %f (requested: %f)" % (self.MAX_QUIESCE_FACTOR, quiesce_factor))
+            quiesce_factor = self.MAX_QUIESCE_FACTOR
+
+        if quiesce_factor < self.MIN_QUIESCE_FACTOR:
+            self.logger.warn("setting the quiesce factor to %f (requested: %f)" % (self.MIN_QUIESCE_FACTOR, quiesce_factor))
+            quiesce_factor = self.MIN_QUIESCE_FACTOR
+
+
+        self.quiesce_factor = max(0.01, quiesce_factor)
+        self.min_quiesce = min_quiesce
+        self.max_quiesce = max_quiesce
+        self.initial_delay = initial_delay
+
+    def next_quiesce_duration(self):
+        mu = (self.min_quiesce + self.max_quiesce) / 2
+        sigma = 3 * math.sqrt(self.max_quiesce - self.min_quiesce)
+        duration = round(self.rnd.gauss(mu, sigma), 1)
+        duration = max(duration, self.min_quiesce)
+        duration = min(duration, self.max_quiesce)
+        return duration
+
+    def tell_quiesce_leader(self, *args):
+        leader = None
+        rc = None
+        stdout = None
+
+        while leader is None and not self.is_stopped:
+            leader = self.fs.get_var('qdb_leader')
```

Comment:
This does an implicit `fs dump`. I think it would be appropriate to get the `FSMap` once when attempting a quiesce and use it for the entire operation. If the MDS cluster changes at any point (failover), abort the whole thing with an explanation.

Reply:
I think that this approach will greatly limit the usability of this thrasher. We may want to run it along with failover scenarios, since the CANCELed state is not considered an error following the discussion we had about that.

Why are you concerned with the `fs dump` once in several minutes? It is just printing the known fs map, isn't it? Is there another way of getting an up to date fsmap from python?

Reply:
As long as `fs dump` is not too frequent (for performance reasons), it's not a big deal.

My concern is mostly that the thrasher would continue working during a failover when perhaps it should stop. If you are already handling it, then ok.

Reply:
Following our slack conversation, there aren't too many options here. We shouldn't be concerned about the overhead of this, as it's a single additional request per quiesce which shouldn't make much difference.

In the future, if this becomes an issue, we can do one of the following, or find another solution:
* cache the leader and only request a new one if there's been a failover. This may be tricky since it will be hard to tell a quiesce error from the "daemon not found" error
* change to calling the mgr `fs quiesce` API. For this, we'll have to list all subvolumes in the request, and that itself may be a separate call unless we add a new mode to the `fs quiesce` call to quiesce all subvolumes of a group or all subvolumes